types/src/token/account.rs

```
@@ -0,0 +1,169 @@
+                lhs.cmp(&rhs)
+            }
+            (Contract(lhs), Contract(rhs)) => lhs.cmp(rhs),
+            // An externally owned account is defined as always "smaller" than a
+            // contract account. This ensures they are never mixed
+            // when ordering.
+            (External(_lhs), Contract(_rhs)) => Ordering::Greater,
+            (Contract(_lhs), External(_rhs)) => Ordering::Less,
+        }
+    }
+}
+
+/// The data an account has in the contract.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Archive, Serialize, Deserialize)]
+#[archive_attr(derive(CheckBytes))]
+pub struct AccountInfo {
+    /// The balance of the account.
+    pub balance: u64,
+    /// The current nonce of the account. Use the current value +1 to perform
+    /// an interaction with the account.
+    pub nonce: u64,
+    /// Status of the account.
+    ///
+    /// # Variants
+    /// 0: No Status
+    /// 1: Frozen
+    /// 2: Blocked
+    // TODO: We want to have this as a `Role` enum soon, but for serialization we use u64 temporarily.
+    pub status: u64,
+}
+
+impl AccountInfo {
+    const FROZEN: u64 = 1;
+    const BLOCKED: u64 = 2;
+
+    /// An empty account.
+    pub const EMPTY: Self = Self {
+        balance: 0,
+        nonce: 0,
+        status: 0,
+    };
+
+    /// Check if the account is blocked.
+    pub fn is_blocked(&self) -> bool {
+        self.status == Self::BLOCKED
+    }
+
+    /// Check if the account is frozen.
+    pub fn is_frozen(&self) -> bool {
+        self.status == Self::FROZEN
+    }
+
+    /// Freeze the account.
+    pub fn freeze(&mut self) {
+        self.status = Self::FROZEN;
+    }
+
+    /// Block the account.
+    pub fn block(&mut self) {
+        self.status = Self::BLOCKED;
+    }
+
+    /// Unfreeze the account.
+    pub fn unfreeze(&mut self) {
+        self.status = 0;
+    }
+
+    /// Unblock the account.
+    pub fn unblock(&mut self) {
+        self.status = 0;
+    }
```

Didn't we want to add a check to make sure the account is actually `frozen` or `blocked` when respectively `unfreezing` or `unblocking`?

---

Yes, and that does happen, but not in the setter here. The actual contract code ensures that the function is not called in the wrong scenario.

Theoretically, we could also make the setter return a result here & let the contract panic via .expect()

---

it would ensure this functionality on a deeper level

---

https://github.com/dusk-network/electronic-money-token/issues/4